File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int len = nums.size();
        vector<int> rpr(len);

        int currPr = 1;
        for (int i = len - 1; i >= 0; i--) {
            currPr *= nums[i];

            rpr[i] = currPr;
        }

        currPr = 1;
        vector<int> res(len);
        for (int i = 0; i < len - 1; i++) {
            res[i] = currPr * rpr[i + 1];

            currPr *= nums[i];
        }
        res[len - 1] = currPr;

        return res;
        
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int product = 1;
        int zeroCount = 0;

        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];

            if (num == 0) {
                zeroCount += 1;
            } else {
                product *= num;
            }
        }

        int[] ans = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            if (zeroCount > 1) {
                ans[i] = 0;
                continue;
            }

            if (nums[i] == 0) {
                ans[i] = product;
            } else {
                if (zeroCount == 1) {
                    ans[i] = 0;
                } else {
                    ans[i] = product / nums[i];
                }
            }
        }

        return ans;
    }
}